var/spack/repos/builtin/packages/p7zip/gcc10.patch

```diff
@@ -0,0 +1,13 @@
+diff --git a/CPP/Windows/ErrorMsg.cpp.orig b/CPP/Windows/ErrorMsg.cpp
+index 99684ae..68416d6 100644
+--- a/CPP/Windows/ErrorMsg.cpp.orig
++++ b/CPP/Windows/ErrorMsg.cpp
+@@ -13,7 +13,7 @@ UString MyFormatMessage(DWORD errorCode)
+   const char * txt = 0;
+   AString msg;
+ 
+-  switch(errorCode) {
++  switch((Int32)errorCode) {
```

Comment:
According https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types , it should be UInt32.

Reply:
I tried it and get an error, on Ubuntu 21.
Does it work for you ?
```
==> p7zip: Executing phase: 'edit'
==> p7zip: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j2' 'all3'

16 errors found in build log:
     173    g++ -c -I. -I../../../../C -I../../../../CPP/myWindows -I../../../../CPP/include_wi
            ndows -I../../../../CPP -O -s -pipe -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DND
            EBUG -D_REENTRANT -DENV_UNIX -D_7ZIP_LARGE_PAGES  -D_FILE_OFFSET_BITS=64 -D_LARGEFI
            LE_SOURCE -D_REENTRANT -DENV_UNIX -DBREAK_HANDLER -DUNICODE -D_UNICODE -DUNIX_USE_W
            IN_FILE   ../../../../CPP/7zip/Archive/Zip/ZipAddCommon.cpp
     174    g++ -c -I. -I../../../../C -I../../../../CPP/myWindows -I../../../../CPP/include_wi
            ndows -I../../../../CPP -O -s -pipe -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DND
            EBUG -D_REENTRANT -DENV_UNIX -D_7ZIP_LARGE_PAGES  -D_FILE_OFFSET_BITS=64 -D_LARGEFI
            LE_SOURCE -D_REENTRANT -DENV_UNIX -DBREAK_HANDLER -DUNICODE -D_UNICODE -DUNIX_USE_W
            IN_FILE -DEXTRACT_ONLY -DNO_READ_FROM_CODER -D_SFX   ../../../../CPP/Windows/ErrorM
            sg.cpp
     175    In file included from ../../../../CPP/Common/Common.h:8,
     176                     from ../../../../CPP/myWindows/StdAfx.h:17,
     177                     from ../../../../CPP/Windows/ErrorMsg.cpp:3:
     178    ../../../../CPP/Windows/ErrorMsg.cpp: In function 'UString NWindows::NError::MyForm
            atMessage(DWORD)':
  >> 179    ../../../../CPP/Common/MyWindows.h:83:40: error: narrowing conversion of '-21474672
            63' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     180       83 | #define E_NOTIMPL ((HRESULT)0x80004001L)
     181          |                                        ^
     182    ../../../../CPP/Windows/ErrorMsg.cpp:18:10: note: in expansion of macro 'E_NOTIMPL'
     183       18 |     case E_NOTIMPL             : txt = "E_NOTIMPL"; break ;
     184          |          ^~~~~~~~~
  >> 185    ../../../../CPP/Common/MyWindows.h:84:44: error: narrowing conversion of '-21474672
            62' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     186       84 | #define E_NOINTERFACE ((HRESULT)0x80004002L)
     187          |                                            ^
     188    ../../../../CPP/Windows/ErrorMsg.cpp:19:10: note: in expansion of macro 'E_NOINTERF
            ACE'
     189       19 |     case E_NOINTERFACE         : txt = "E_NOINTERFACE"; break ;
     190          |          ^~~~~~~~~~~~~
  >> 191    ../../../../CPP/Common/MyWindows.h:85:38: error: narrowing conversion of '-21474672
            60' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     192       85 | #define E_ABORT ((HRESULT)0x80004004L)
     193          |                                      ^
     194    ../../../../CPP/Windows/ErrorMsg.cpp:20:10: note: in expansion of macro 'E_ABORT'
     195       20 |     case E_ABORT               : txt = "E_ABORT"; break ;
     196          |          ^~~~~~~
  >> 197    ../../../../CPP/Common/MyWindows.h:86:37: error: narrowing conversion of '-21474672
            59' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     198       86 | #define E_FAIL ((HRESULT)0x80004005L)
     199          |                                     ^
     200    ../../../../CPP/Windows/ErrorMsg.cpp:21:10: note: in expansion of macro 'E_FAIL'
     201       21 |     case E_FAIL                : txt = "E_FAIL"; break ;
     202          |          ^~~~~~
  >> 203    ../../../../CPP/Common/MyWindows.h:87:52: error: narrowing conversion of '-21472870
            39' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     204       87 | #define STG_E_INVALIDFUNCTION ((HRESULT)0x80030001L)
     205          |                                                    ^
     206    ../../../../CPP/Windows/ErrorMsg.cpp:22:10: note: in expansion of macro 'STG_E_INVA
            LIDFUNCTION'
     207       22 |     case STG_E_INVALIDFUNCTION : txt = "STG_E_INVALIDFUNCTION"; break ;
     208          |          ^~~~~~~~~~~~~~~~~~~~~
  >> 209    ../../../../CPP/Common/MyWindows.h:88:44: error: narrowing conversion of '-21470248
            82' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     210       88 | #define E_OUTOFMEMORY ((HRESULT)0x8007000EL)
     211          |                                            ^
     212    ../../../../CPP/Windows/ErrorMsg.cpp:23:10: note: in expansion of macro 'E_OUTOFMEM
            ORY'
     213       23 |     case E_OUTOFMEMORY         : txt = "E_OUTOFMEMORY"; break ;
     214          |          ^~~~~~~~~~~~~
  >> 215    ../../../../CPP/Common/MyWindows.h:89:43: error: narrowing conversion of '-21470248
            09' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     216       89 | #define E_INVALIDARG ((HRESULT)0x80070057L)
     217          |                                           ^
     218    ../../../../CPP/Windows/ErrorMsg.cpp:24:10: note: in expansion of macro 'E_INVALIDA
            RG'
     219       24 |     case E_INVALIDARG          : txt = "E_INVALIDARG"; break ;
     220          |          ^~~~~~~~~~~~
  >> 221    make[1]: *** [makefile.list:290: ErrorMsg.o] Error 1
     222    make[1]: Leaving directory '/tmp/cessenat/spack-stage/spack-stage-p7zip-16.02-aomwh
            ygxgbqe6z5uapmdbo5ydpu34oxo/spack-src/CPP/7zip/Bundles/SFXCon'
     223    make: *** [makefile:40: sfx] Error 2
     224    make: *** Waiting for unfinished jobs....
     225    g++ -c -I. -I../../../../C -I../../../../CPP/myWindows -I../../../../CPP/include_wi
            ndows -I../../../../CPP -O -s -pipe -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DND
            EBUG -D_REENTRANT -DENV_UNIX -D_7ZIP_LARGE_PAGES  -D_FILE_OFFSET_BITS=64 -D_LARGEFI
            LE_SOURCE -D_REENTRANT -DENV_UNIX -DBREAK_HANDLER -DUNICODE -D_UNICODE -DUNIX_USE_W
            IN_FILE   ../../../../CPP/7zip/Archive/Zip/ZipHandler.cpp
     226    g++ -c -I. -I../../../../C -I../../../../CPP/myWindows -I../../../../CPP/include_wi
            ndows -I../../../../CPP -O -s -pipe -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DND
            EBUG -D_REENTRANT -DENV_UNIX -D_7ZIP_LARGE_PAGES  -D_FILE_OFFSET_BITS=64 -D_LARGEFI
            LE_SOURCE -D_REENTRANT -DENV_UNIX -DBREAK_HANDLER -DUNICODE -D_UNICODE -DUNIX_USE_W
            IN_FILE   ../../../../CPP/7zip/Archive/Zip/ZipHandlerOut.cpp
     227    g++ -c -I. -I../../../../C -I../../../../CPP/myWindows -I../../../../CPP/include_wi
            ndows -I../../../../CPP -O -s -pipe -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DND
            EBUG -D_REENTRANT -DENV_UNIX -D_7ZIP_LARGE_PAGES  -D_FILE_OFFSET_BITS=64 -D_LARGEFI
            LE_SOURCE -D_REENTRANT -DENV_UNIX -DBREAK_HANDLER -DUNICODE -D_UNICODE -DUNIX_USE_W
            IN_FILE   ../../../../CPP/7zip/Archive/Zip/ZipIn.cpp

     ...

     347    g++ -c -I. -I../../../../C -I../../../../CPP/myWindows -I../../../../CPP/include_wi
            ndows -I../../../../CPP -O -s -pipe -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DND
            EBUG -D_REENTRANT -DENV_UNIX -D_7ZIP_LARGE_PAGES  -D_FILE_OFFSET_BITS=64 -D_LARGEFI
            LE_SOURCE -D_REENTRANT -DENV_UNIX -DBREAK_HANDLER -DUNICODE -D_UNICODE -DUNIX_USE_W
            IN_FILE   ../../../../CPP/Common/XzCrc64Reg.cpp
     348    g++ -c -I. -I../../../../C -I../../../../CPP/myWindows -I../../../../CPP/include_wi
            ndows -I../../../../CPP -O -s -pipe -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DND
            EBUG -D_REENTRANT -DENV_UNIX -D_7ZIP_LARGE_PAGES  -D_FILE_OFFSET_BITS=64 -D_LARGEFI
            LE_SOURCE -D_REENTRANT -DENV_UNIX -DBREAK_HANDLER -DUNICODE -D_UNICODE -DUNIX_USE_W
            IN_FILE   ../../../../CPP/Windows/ErrorMsg.cpp
     349    In file included from ../../../../CPP/Common/Common.h:8,
     350                     from ../../../../CPP/myWindows/StdAfx.h:17,
     351                     from ../../../../CPP/Windows/ErrorMsg.cpp:3:
     352    ../../../../CPP/Windows/ErrorMsg.cpp: In function 'UString NWindows::NError::MyForm
            atMessage(DWORD)':
  >> 353    ../../../../CPP/Common/MyWindows.h:83:40: error: narrowing conversion of '-21474672
            63' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     354       83 | #define E_NOTIMPL ((HRESULT)0x80004001L)
     355          |                                        ^
     356    ../../../../CPP/Windows/ErrorMsg.cpp:18:10: note: in expansion of macro 'E_NOTIMPL'
     357       18 |     case E_NOTIMPL             : txt = "E_NOTIMPL"; break ;
     358          |          ^~~~~~~~~
  >> 359    ../../../../CPP/Common/MyWindows.h:84:44: error: narrowing conversion of '-21474672
            62' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     360       84 | #define E_NOINTERFACE ((HRESULT)0x80004002L)
     361          |                                            ^
     362    ../../../../CPP/Windows/ErrorMsg.cpp:19:10: note: in expansion of macro 'E_NOINTERF
            ACE'
     363       19 |     case E_NOINTERFACE         : txt = "E_NOINTERFACE"; break ;
     364          |          ^~~~~~~~~~~~~
  >> 365    ../../../../CPP/Common/MyWindows.h:85:38: error: narrowing conversion of '-21474672
            60' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     366       85 | #define E_ABORT ((HRESULT)0x80004004L)
     367          |                                      ^
     368    ../../../../CPP/Windows/ErrorMsg.cpp:20:10: note: in expansion of macro 'E_ABORT'
     369       20 |     case E_ABORT               : txt = "E_ABORT"; break ;
     370          |          ^~~~~~~
  >> 371    ../../../../CPP/Common/MyWindows.h:86:37: error: narrowing conversion of '-21474672
            59' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     372       86 | #define E_FAIL ((HRESULT)0x80004005L)
     373          |                                     ^
     374    ../../../../CPP/Windows/ErrorMsg.cpp:21:10: note: in expansion of macro 'E_FAIL'
     375       21 |     case E_FAIL                : txt = "E_FAIL"; break ;
     376          |          ^~~~~~
  >> 377    ../../../../CPP/Common/MyWindows.h:87:52: error: narrowing conversion of '-21472870
            39' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     378       87 | #define STG_E_INVALIDFUNCTION ((HRESULT)0x80030001L)
     379          |                                                    ^
     380    ../../../../CPP/Windows/ErrorMsg.cpp:22:10: note: in expansion of macro 'STG_E_INVA
            LIDFUNCTION'
     381       22 |     case STG_E_INVALIDFUNCTION : txt = "STG_E_INVALIDFUNCTION"; break ;
     382          |          ^~~~~~~~~~~~~~~~~~~~~
  >> 383    ../../../../CPP/Common/MyWindows.h:88:44: error: narrowing conversion of '-21470248
            82' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     384       88 | #define E_OUTOFMEMORY ((HRESULT)0x8007000EL)
     385          |                                            ^
     386    ../../../../CPP/Windows/ErrorMsg.cpp:23:10: note: in expansion of macro 'E_OUTOFMEM
            ORY'
     387       23 |     case E_OUTOFMEMORY         : txt = "E_OUTOFMEMORY"; break ;
     388          |          ^~~~~~~~~~~~~
  >> 389    ../../../../CPP/Common/MyWindows.h:89:43: error: narrowing conversion of '-21470248
            09' from 'LONG' {aka 'int'} to 'unsigned int' [-Wnarrowing]
     390       89 | #define E_INVALIDARG ((HRESULT)0x80070057L)
     391          |                                           ^
     392    ../../../../CPP/Windows/ErrorMsg.cpp:24:10: note: in expansion of macro 'E_INVALIDA
            RG'
     393       24 |     case E_INVALIDARG          : txt = "E_INVALIDARG"; break ;
     394          |          ^~~~~~~~~~~~
  >> 395    make[1]: *** [makefile.list:652: ErrorMsg.o] Error 1
     396    make[1]: *** Waiting for unfinished jobs....
     397    make[1]: Leaving directory '/tmp/cessenat/spack-stage/spack-stage-p7zip-16.02-aomwh
            ygxgbqe6z5uapmdbo5ydpu34oxo/spack-src/CPP/7zip/Bundles/Alone'
     398    make: *** [makefile:17: 7za] Error 2

See build log for details:
  /tmp/cessenat/spack-stage/spack-stage-p7zip-16.02-aomwhygxgbqe6z5uapmdbo5ydpu34oxo/spack-build-out.txt

```

Reply:
BTW when applying E. Brugger's patch I checked 7z x worked on 
https://downloads.sourceforge.net/project/sevenzip/7-Zip/21.02/7z2102-src.7z
I could not make a recipe for spack since 7z in not is spack pre-requisites.
I would be glad if you had an idea for a work around.

Reply:
I am new to spack packaging as well. I was thinking of adding a custom package in line with Intel compiler or other pre-built packages for 7zip and build from sources using that bootstrap if explicit architecture or optimization level is requested.
----
I guess DWORD is declared somewhere else in code as long and I think your original changes was good enough for now :). Sorry for the trouble.

Reply:
Thanks for the idea.
I'm afraid you still have to approve my PR, don't you ?